sound/components/virt.c

```diff
@@ -0,0 +1,248 @@
+static bool started;
+
+static void respond_to_cmd(sound_queues_t *client_rings,
+                           sound_cmd_t *cmd,
+                           sound_status_t status)
+{
+    cmd->status = status;
+    if (sound_enqueue_cmd(client_rings->cmd_res, cmd) != 0) {
+        microkit_dbg_puts("SND VIRT|ERR: failed to respond to command\n");
+    }
+}
+
+static void respond_to_pcm(sound_queues_t *client_rings,
+                           sound_pcm_t *pcm,
+                           sound_status_t status)
+{
+    pcm->status = status;
+    pcm->latency_bytes = 0;
+    if (sound_enqueue_pcm(client_rings->pcm_res, pcm) != 0) {
+        microkit_dbg_puts("SND VIRT|ERR: failed to respond to pcm\n");
+    }
+}
+
+static int notified_by_client(int client)
+{
+    if (client < 0 || client > CLIENT_COUNT) {
+        microkit_dbg_puts("SND VIRT|ERR: invalid client id\n");
+        return -1;
+    }
+
+    bool notify_client = false;
+    bool notify_driver = false;
+    sound_queues_t *client_rings = &clients[client];
+    sound_cmd_t cmd;
+
+    while (sound_dequeue_cmd(client_rings->cmd_req, &cmd) == 0) {
+
+        if (cmd.stream_id > MAX_STREAMS) {
+            microkit_dbg_puts("SND VIRT|ERR: stream id too large\n");
+            respond_to_cmd(client_rings, &cmd, SOUND_S_BAD_MSG);
+            continue;
+        }
+
+        if (owners[cmd.stream_id] == -1) {
+            if (cmd.code == SOUND_CMD_TAKE) {
+                owners[cmd.stream_id] = client;
+            } else {
+                microkit_dbg_puts("SND VIRT|ERR: client must take first\n");
+                respond_to_cmd(client_rings, &cmd, SOUND_S_BAD_MSG);
+                notify_client = true;
+                continue;
+            }
+        }
+
+        int owner = owners[cmd.stream_id];
+        if (owner != client) {
+            microkit_dbg_puts("SND VIRT|ERR: stream busy\n");
+            respond_to_cmd(client_rings, &cmd, SOUND_S_BUSY);
+            notify_client = true;
+            continue;
+        }
+
+        if (sound_enqueue_cmd(driver_rings.cmd_req, &cmd) != 0) {
+            microkit_dbg_puts("SND VIRT|ERR: Failed to enqueue command\n");
+            return -1;
+        }
+        notify_driver = true;
+    }
+
+    sound_pcm_t pcm;
+    while (sound_dequeue_pcm(client_rings->pcm_req, &pcm) == 0) {
+
+        if (pcm.stream_id > MAX_STREAMS) {
+            microkit_dbg_puts("SND VIRT|ERR: stream id too large\n");
+            respond_to_pcm(client_rings, &pcm, SOUND_S_BAD_MSG);
+            notify_client = true;
+            continue;
+        }
+
+        int owner = owners[pcm.stream_id];
+        if (owner != client) {
+            microkit_dbg_puts("SND VIRT|ERR: driver replied to busy stream\n");
+            respond_to_pcm(client_rings, &pcm, SOUND_S_BAD_MSG);
+            notify_client = true;
+            continue;
+        }
+
+        // Write PCM data to RAM
+        microkit_arm_vspace_data_clean(pcm.addr, pcm.addr + pcm.len);
```

Comment:
```suggestion
        cache_clean(pcm.addr, pcm.addr + pcm.len);
```

Unlike invalidate, we can use user-space operations which is available from `cache.c`.